contracts/staking/ValidatorManager.sol

```diff
@@ -93,6 +90,105 @@ abstract contract ValidatorManager is
+    }
+
+    modifier initializedValidatorSet() {
+        require(
+            _getValidatorManagerStorage()._initializedValidatorSet,
+            "ValidatorManager: validator set not initialized"
+        );
+        _;
+    }
+
+    /**
+     * @notice See {IValidatorManager-initializeValidatorSet}.
+     */
+    function initializeValidatorSet(
+        SubnetConversionData calldata subnetConversionData,
+        uint32 messageIndex
+    ) external {
+        ValidatorManagerStorage storage $ = _getValidatorManagerStorage();
+        require(!$._initializedValidatorSet, "ValidatorManager: already initialized validator set");
+        require(
+            subnetConversionData.blockchainID == WARP_MESSENGER.getBlockchainID(),
+            "ValidatorManager: invalid blockchain ID"
+        );
+        require(
+            address(subnetConversionData.validatorManagerAddress) == address(this),
+            "ValidatorManager: invalid validator manager address"
+        );
+        require(
+            subnetConversionData.initialValidators.length > 0,
+            "ValidatorManager: no initial validators"
+        );
+        uint256 numInitialValidators = subnetConversionData.initialValidators.length;
+        require(
+            numInitialValidators < type(uint32).max, "ValidatorManager: too many initial validators"
+        );
+
+        // Verify that the sha256 hash of the Subnet conversion data matches with the Warp message's subnetConversionID.
+        bytes memory encodedConversion = abi.encodePacked(
+            subnetConversionData.convertSubnetTxID,
+            subnetConversionData.blockchainID,
+            uint32(20),
+            subnetConversionData.validatorManagerAddress,
+            uint32(numInitialValidators)
+        );
+        for (uint256 i; i < numInitialValidators; i++) {
+            InitialValidator memory initialValidator = subnetConversionData.initialValidators[i];
+            bytes32 nodeID = initialValidator.nodeID;
+            require(nodeID != bytes32(0), "ValidatorManager: invalid node ID");
+            require(
+                $._activeValidators[nodeID] == bytes32(0),
+                "ValidatorManager: node ID already active"
+            );
+            require(
+                initialValidator.blsPublicKey.length == 48,
+                "ValidatorManager: invalid blsPublicKey length"
+            );
+
+            // Continue to encode the initial validators.
+            encodedConversion = abi.encodePacked(
+                encodedConversion,
+                initialValidator.nodeID,
+                initialValidator.weight,
+                initialValidator.blsPublicKey
+            );
+
+            // Validation ID of the initial validators is the sha256 hash of the
+            // convert Subnet tx ID and the index of the initial validator.
+            bytes32 validationID =
+                sha256(abi.encodePacked(subnetConversionData.convertSubnetTxID, i));
```

Comment:
We'll need to make sure `i` here is represented as 4 bytes (uint32) in the pre-image. I'm not sure if `encodePacked` would handle that properly if we use a `uint32` instead of ` uint256`, or if we have to pack it manually.

Reply:
Using `uint32` as iterator is sufficient. I made the change in #553